bitcoin/crawler.go

```diff
@@ -0,0 +1,539 @@
+							maStr := fmt.Sprintf("/ip4/%s/tcp/%d", addr.IP.String(), addr.Port)
+							maddr, err := ma.NewMultiaddr(maStr)
+							if err != nil {
+								continue // Skip invalid addresses
+							}
+
+							mapped[i] = PeerInfo{
+								AddrInfo: AddrInfo{
+									id:   maddr.String(),
+									Addr: []ma.Multiaddr{maddr},
+								},
+							}
+						}
+						return mapped
+					}()...)
+
+					if firstReceived == -1 {
+						firstReceived = len(tmsg.AddrList)
+					} else if firstReceived > len(tmsg.AddrList) || firstReceived == 0 {
+						// Probably done.
+						break Loop
+					}
+				case *wire.MsgPing:
+					log.WithField("addr", pi.Addr).WithField("toleratePingMessages", toleratePingMessages).Debugln("Sending Pong message...")
+					toleratePingMessages--
+					err = c.WriteMessage(conn, wire.NewMsgPong(tmsg.Nonce))
+					if err != nil {
+						log.Errorf("Pong msg err: %s", err)
+						break Loop
+					}
+					if toleratePingMessages < 0 {
+						log.Debugln("Ran out of limit to tolerate Ping messages")
+						break Loop
+					}
+				case *wire.MsgVerAck:
+					tolerateVerAckMessages--
+					if tolerateVerAckMessages < 0 {
+						log.Debugln("Ran out of limit to tolerate Ver Ack messages")
+						break Loop
+					}
+				case *wire.MsgInv:
+					tolerateInvMessages--
+					if tolerateInvMessages < 0 {
+						log.Debugln("Ran out of limit to tolerate Inv messages")
+						break Loop
+					}
+				default:
+					if tmsg != nil {
+						log.WithField("msg_type", tmsg.Command()).Debugf("Found other message from %s", pi.Addr)
+					}
+					tolerateMessages--
+				}
+				err = c.WriteMessage(conn, wire.NewMsgGetAddr())
+				if err != nil {
+					log.Infoln("Error when requesting peers")
+					break Loop
+				}
+
+			}
+
+			if len(neighbours) > 0 {
+				log.WithField("num_peers", len(neighbours)).WithField("addr", pi.Addr).Infoln("Found peers")
+			} else {
+				log.WithField("addr", pi.Addr).Infoln("Found no peers")
+			}
+
+		} else {
+			result.Error = result.ConnectError
+		}
+
+		result.RoutingTable = &core.RoutingTable[PeerInfo]{
+			PeerID:    pi.ID(),
+			Neighbors: neighbours,
+			ErrorBits: uint16(0), // FIXME
+			Error:     result.Error,
+		}
+
+		// if there was a connection error, parse it to a known one
+		if result.ConnectError != nil {
+			result.ConnectErrorStr = db.NetError(result.ConnectError)
+		} else {
+			// Free connection resources
+			if err := conn.Close(); err != nil {
+				log.WithError(err).WithField("remoteID", pi.ID().ShortString()).Warnln("Could not close connection to peer")
+			}
+			conn = nil
+		}
+
+		if result.Error != nil {
+			result.ErrorStr = db.NetError(result.Error)
+		}
+
+		// send the result back and close channel
+		select {
+		case resultCh <- result:
+		case <-ctx.Done():
+			if conn != nil {
+				if err := conn.Close(); err != nil {
+					log.WithError(err).WithField("remoteID", pi.ID().ShortString()).Warnln("Could not close connection on context cancel")
+				}
+			}
```

Comment:
rather put this in a `defer` statement at the point in the code where you have received the connection object.